### PR TITLE
Add microsecond precision support for timestamp

### PR DESCRIPTION
**Pull request by vigneshwarselvaraj**
TL;DR Presto supports Millisecond precision for Timestamp read / write. This change provides an option in presto-orc to enable Timestamp writes and reads with microsecond precision. 

This change is limited only to presto-orc, we use this as a standalone reader / writer library. 

A new timestamp type TIMESTAMP_MICROSECONDS was introduced earlier to handle microsecond precision. This change enables this new type to be used by TimestampColumnWriter to write with either microsecond precision. This allows the existing type TIMESTAMP to continue to work with millisecond precision by presto.

OrcSelectiveRecordReader is already able to support reading timestamps in microsecond precision if enableTimestampMicroPrecision is set as true in OrcReaderOptions. Validated this read works as expected.

Test plan:

- Added new unit tests to validate changes
- Tested in Validation service for about 4 hours, no issues found

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Review by vigneshwarselvaraj — COMMENTED** [review]
@arunthirupathi @sdruzkin Please help review

**Comment by sdruzkin** [comment]
Can you force push this commit again to re-trigger tests?
Also please do sign EasyCLA.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcWriter.java [hunk 1]
```diff
@@ -338,6 +344,13 @@ public OrcWriter(
         stats.updateSizeInBytes(previouslyRecordedSizeInBytes);
     }
 
+    private void enableTimestampMicroPrecision()
+    {
+        if (!this.enableTimestampMicroPrecision) {
```
Comment (sdruzkin) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcWriter.java:349]: No need to check this condition.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcWriter.java [hunk 2]
```diff
@@ -838,7 +851,8 @@ public void validate(OrcDataSource input)
                 types,
                 hiveStorageTimeZone,
                 orcEncoding,
-                new OrcReaderOptions(new DataSize(1, MEGABYTE), new DataSize(8, MEGABYTE), new DataSize(16, MEGABYTE), false),
+                new OrcReaderOptions(new DataSize(1, MEGABYTE), new DataSize(8, MEGABYTE), new DataSize(16, MEGABYTE), false,
```
Comment (sdruzkin) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcWriter.java:854]: Reformat as:
```
new OrcReaderOptions(
  arg1,
  arg2,
  argN);
```

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/TimestampColumnWriter.java [hunk 3]
```diff
@@ -171,7 +188,7 @@ public long writeBlock(Block block)
                 // In Presto, we only have millisecond precision.
```
Comment (sdruzkin) [inline presto-orc/src/main/java/com/facebook/presto/orc/writer/TimestampColumnWriter.java:188]:
```
In Presto, we only have millisecond precision.
```
Doesn't seem to be true anymore.

File: presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java [hunk 4]
```diff
@@ -2210,6 +2240,13 @@ else if (type.equals(TIMESTAMP)) {
             long millisUtc = (int) ((SqlTimestamp) value).getMillisUtc();
             return new Timestamp(millisUtc);
         }
+        else if (type.equals(TIMESTAMP_MICROSECONDS)) {
+            long microsUtc = ((SqlTimestamp) value).getMicrosUtc();
+            int nanos = (int) (microsUtc % 1_000_000) * 1000;
```
Comment (sdruzkin) [inline presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java:2245]: Cast using Math.toIntExact

File: presto-orc/src/test/java/com/facebook/presto/orc/TestTimestampWriteAndRead.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+import static com.facebook.presto.testing.DateTimeTestingUtils.sqlTimestampOf;
+import static com.facebook.presto.testing.TestingConnectorSession.SESSION;
+import static java.lang.Math.floorDiv;
+import static java.util.concurrent.TimeUnit.MICROSECONDS;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.stream.Collectors.toList;
+
+public class TestTimestampWriteAndRead
+{
+    private static final List<SqlTimestamp> MICRO_TIMESTAMPS = ImmutableList.of(
```
Comment (sdruzkin) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestTimestampWriteAndRead.java:41]: Can you rename it to MICROSECOND_VALUES to make it more distinguished from TIMESTAMP_MICROSECONDS?

File: presto-orc/src/test/java/com/facebook/presto/orc/TestTimestampWriteAndRead.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+            sqlTimestampOf(0L, SESSION, MICROSECONDS),
+            sqlTimestampOf(1L, SESSION, MICROSECONDS),
+            sqlTimestampOf(999_999L, SESSION, MICROSECONDS),
+            sqlTimestampOf(1_000_000L, SESSION, MICROSECONDS),
+            sqlTimestampOf(-60_000_000_000_000_789L, SESSION, MICROSECONDS),
+            sqlTimestampOf(-230_000_000_000_999_999L, SESSION, MICROSECONDS),
+            sqlTimestampOf(1_650_483_250_000_507L, SESSION, MICROSECONDS),
+            sqlTimestampOf(60_000_000_000_123_789L, SESSION, MICROSECONDS),
+            sqlTimestampOf(230_000_000_000_999_999L, SESSION, MICROSECONDS));
```
Comment (sdruzkin) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestTimestampWriteAndRead.java:50]: What are these values? Just some random values?

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcWriter.java [hunk 7]
```diff
@@ -305,6 +307,10 @@ public OrcWriter(
                     orcEncoding.createMetadataWriter());
             columnWriters.add(columnWriter);
 
+            if (fieldType == TIMESTAMP_MICROSECONDS) {
```
[the remark this answers is not in the document]
Comment (vigneshwarselvaraj) [reply presto-orc/src/main/java/com/facebook/presto/orc/OrcWriter.java:310]: Writing different precisions works fine from writer perspective, but would fail at validate() since OrcReaderOptions support only reading in either Micro or Milli precision. I also believe we should allow writing in a single precision only. I'll look into handling this for nested types.

[the remark this answers is not in the document]
Comment (arunthirupathi) [reply presto-orc/src/main/java/com/facebook/presto/orc/OrcWriter.java:310]: We should remove the enableTimestampMicroPrecision flag and make it purely type driven. 

If writer gives type as
TIMESTAMP -> long is interpreted as millis.
TIMESTAMP_MICROS -> long is interpreted as microseconds.

Similarly on the reader it will be 
When the requested type is TIMESTAMP -> long values are of milliseconds precision.
TIMESTAMP_MICROS -> long values are of microseconds precision.

**Comment by vigneshwarselvaraj** [comment]
Thanks for reviewing the changes !! I'll raise a different set of changes to remove microsecond option and make it type driven.
